internal/api/permissions.go

```diff
@@ -72,6 +90,169 @@ func (r *Router) checkAction(c echo.Context) error {
 	return c.JSON(http.StatusOK, echo.Map{})
 }
 
+type checkPermissionsRequest struct {
+	Actions []checkAction `json:"actions"`
+}
+
+type checkAction struct {
+	ResourceID string `json:"resource_id"`
+	Action     string `json:"action"`
+}
+
+type checkStatus struct {
+	Resource types.Resource
+	Action   string
+	Error    error
+}
+
+// checkAllActions will check if a subject is allowed to perform an action on a list of resources.
+// This is the permissions check endpoint.
+// It will return a 200 if the subject is allowed to perform all requested resource actions.
+// It will return a 400 if the request is invalid.
+// It will return a 403 if the subject is not allowed to perform all requested resource actions.
+//
+// Note that this expects a JWT token to be present in the request. This token must
+// contain the subject of the request in the "sub" claim.
+func (r *Router) checkAllActions(c echo.Context) error {
+	ctx, span := tracer.Start(c.Request().Context(), "api.checkAllActions")
+	defer span.End()
+
+	// Subject validation
+	subject, err := currentSubject(c)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "failed to get the subject").SetInternal(err)
+	}
+
+	subjectResource, err := r.engine.NewResourceFromID(subject)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "error processing subject ID").SetInternal(err)
+	}
+
+	var reqBody checkPermissionsRequest
+
+	if err := c.Bind(&reqBody); err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, "error parsing request body").SetInternal(err)
+	}
+
+	var errs []error
+
+	results := make([]*checkStatus, len(reqBody.Actions))
+
+	for i, check := range reqBody.Actions {
+		if check.Action == "" {
+			errs = append(errs, fmt.Errorf("check %d: %w", i, ErrNoActionDefined))
+
+			continue
+		}
+
+		resourceID, err := gidx.Parse(check.ResourceID)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("check %d: %w: error parsing resource id: %s", i, err, check.ResourceID))
+
+			continue
+		}
+
+		resource, err := r.engine.NewResourceFromID(resourceID)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("check %d: %w: error creating resource from id: %s", i, err, resourceID.String()))
+
+			continue
+		}
+
+		results[i] = &checkStatus{
+			Resource: resource,
+			Action:   check.Action,
+		}
+	}
+
+	if len(errs) != 0 {
+		return echo.NewHTTPError(http.StatusBadRequest, "invalid check request").SetInternal(multierr.Combine(errs...))
+	}
+
+	checkCh := make(chan *checkStatus)
+
+	wg := new(sync.WaitGroup)
```

Comment:
Rather than use a `sync.WaitGroup` here, we can instead provide a channel for workers to send results to and iterate over that until we get back as many results as we expect, then close the channel to free resources.

Additionally, we can set up a new context using [`context.WithCancelCause`](https://pkg.go.dev/context#WithCancelCause), then when receiving results cancel the entire operation if we get any errors. SpiceDB seems pretty good at handling context propagation and terminating operations early so we can have things fail faster that way.

Reply:
Removed WaitGroup and handle waiting with channels.